src/VisualStudio/Core/Def/Telemetry/VisualStudioTelemetryInitializer.cs

```diff
@@ -0,0 +1,89 @@
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Telemetry;
+using Microsoft.VisualStudio.Telemetry;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServices.Telemetry
+{
+    [ExportEventListener(WellKnownEventListeners.Workspace, WorkspaceKind.Host), Shared]
+    internal class VisualStudioTelemetryInitializer : IEventListener<object>
+    {
+        private readonly VisualStudioWorkspace _visualStudioWorkspace;
+        private readonly IThreadingContext _threadingContext;
+        private readonly IAsynchronousOperationListener _asynchronousOperationListener;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioTelemetryInitializer(
+            VisualStudioWorkspace visualStudioWorkspace,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListenerProvider listenerProvider)
+        {
+            _visualStudioWorkspace = visualStudioWorkspace;
+            _threadingContext = threadingContext;
+            _asynchronousOperationListener = listenerProvider.GetListener(FeatureAttribute.Telemetry);
+        }
+
+        public void StartListening(Workspace workspace, object serviceOpt)
+        {
+            if (workspace is VisualStudioWorkspace)
+            {
+                var _ = StartListeningAsync();
+            }
+        }
+
+        private async Task StartListeningAsync()
+        {
+            // Have to catch all exceptions coming through here as this is called from a
+            // fire-and-forget method and we want to make sure nothing leaks out.
+            try
+            {
+                using var token = _asynchronousOperationListener.BeginAsyncOperation(nameof(InitializeTelemetryAsync));
+                await InitializeTelemetryAsync(_threadingContext.DisposalToken).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                // Cancellation is normal (during VS closing).  Just ignore.
+            }
+            catch (Exception e) when (FatalError.ReportAndCatch(e))
+            {
+                // Otherwise report a watson for any other exception.  Don't bring down VS.  This is
+                // a BG service we don't want impacting the user experience.
+            }
+        }
+
+        private async Task InitializeTelemetryAsync(CancellationToken cancellationToken)
+        {
+            // Fetch the session synchronously on the UI thread; if this doesn't happen before we try using this on
+            // the background thread then we will experience hangs like we see in this bug:
+            // https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?_a=edit&id=190808 or
+            // https://devdiv.visualstudio.com/DevDiv/_workitems?id=296981&_a=edit
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
Per the linked bugs, initializing the session has to happen on the UI thread before we attempt to use it, otherwise we can hang.  This is why I moved the VisualStudioProjectTelemetryService to be triggered explicitly after we wait for the default session here (inside the telemetryService.InitializeTelemetrySessionAsync).

Before, VisualStudioProjectTelemetryService was started on workspace start anyway, so we shouldn't be delaying that in any meaningful way.